sh: fix disassembly of branch opcodes

fenugrec opened the pull request:
The displacement field must be sign-extended and multiplied by 2, not
cast to an unsigned type.
This fixes #9235 . To be clear : issue is a regression caused by part of df82290  , but that commit was partly good. The test suite of https://github.com/radare/radare2-regressions/pull/1148/commits covers both issues.

radare commented:
oh! is this bug fixed in binutils? is this a backport? maybe good to update that .c from latest binutils maybe?

fenugrec commented:
well, binutils' version of these files has diverged somewhat since radare took a copy. But binutils has an explicit cast to "char" for the line I fixed, and nothing at all for the other one. I would say that binutils isn't strictly correct: some of these opcodes (add, bt/bf, etc) have a sign-extended 8-bit immediate (like the one I fixed), and others (and, or, tst, etc) are unsigned 8-bit (like the one df82290 fixed).

radare commented:
So what do u propose?

> On 25 Jan 2018, at 16:47, fenugrec <notifications@github.com> wrote:
> 
> well, binutils' version of these files has diverged somewhat since radare took a copy. But binutils has an explicit cast to "char" for the line I fixed, and nothing at all for the other one. I would say that binutils isn't strictly correct: some of these opcodes (add, bt/bf, etc) have a sign-extended 8-bit immediate (like the one I fixed), and others (and, or, tst, etc) are unsigned 8-bit (like the one df82290 fixed).
> 
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

fenugrec commented:
I suggest
- merging the regression test https://github.com/radare/radare2-regressions/pull/1148 to have some common reference if it breaks again
- filing an issue / PR to upstream binutils. I don't have time to fix their code but I could file an issue
- backport their fix back into r2 if possible

fenugrec commented:
Concidence : somebody already filed a bug report upstream 2 weeks ago  ! https://sourceware.org/bugzilla/show_bug.cgi?id=22699
